File: backend/src/modules/placements/evaluation.service.ts
import { AppError } from "../../shared/app-error.js";
import { InternshipEvaluationRepository } from "./evaluation.repository.js";
import type {
  InternshipEvaluationActor,
  InternshipEvaluationInput,
  InternshipEvaluationRole,
} from "./evaluation.types.js";
import type { RequestMetadata } from "./placement.types.js";

function notFound() {
  return new AppError(404, "INTERNSHIP_PLACEMENT_NOT_FOUND", "Không tìm thấy kỳ thực tập.");
}

export class InternshipEvaluationService {
  constructor(private readonly repository: InternshipEvaluationRepository) {}

  async get(
    actor: InternshipEvaluationActor,
    applicationId: string,
    role: InternshipEvaluationRole,
  ) {
    const context = await this.repository.findContextByApplication(applicationId);
    this.assertOwnership(context, actor, role);
    return this.repository.viewFor(context!, role);
  }

  async submit(
    actor: InternshipEvaluationActor,
    applicationId: string,
    commandId: string,
    role: InternshipEvaluationRole,
    input: InternshipEvaluationInput,
    request: RequestMetadata,
  ) {
    return this.repository.withTransaction(async (client) => {
      const context = await this.repository.findContextByApplication(applicationId, client, true);
      this.assertOwnership(context, actor, role);

      const repeated = await this.repository.findByCommand(context!.placement_id, commandId, client);
      if (repeated) {
        if (repeated.respondentRole !== role) {
          throw new AppError(409, "IDEMPOTENCY_KEY_REUSED", "Idempotency-Key đã được dùng cho một phiếu khác.");
        }
        return this.repository.viewFor(context!, role, client);
      }

      const current = await this.repository.viewFor(context!, role, client);
      if (!current.canSubmit) {
        if (context!.placement_status !== "COMPLETED") {
          throw new AppError(
            409,
            "INTERNSHIP_EVALUATION_NOT_AVAILABLE",
            "Chỉ có thể gửi đánh giá sau khi UIT xác nhận hoàn thành kỳ thực tập.",
          );
        }
        throw new AppError(
          409,
          "INTERNSHIP_EVALUATION_ALREADY_SUBMITTED",
          "Phiếu đánh giá đã được gửi và không thể chỉnh sửa.",
        );
      }

      await this.repository.insert(client, { actor, applicationId, commandId, role, input, request }, context!);
      return this.repository.viewFor(context!, role, client);
    });
  }

  private assertOwnership(
    context: Awaited<ReturnType<InternshipEvaluationRepository["findContextByApplication"]>>,
    actor: InternshipEvaluationActor,
    role: InternshipEvaluationRole,
  ): asserts context {
    if (!context) throw notFound();
    if (role === "STUDENT" && context.student_profile_id !== actor.studentProfileId) throw notFound();
    if (role === "COMPANY" && context.company_id !== actor.companyId) throw notFound();
  }
}

File: backend/src/main/java/vn/edu/uit/careerhub/placements/EvaluationService.java
package vn.edu.uit.careerhub.placements;

import java.util.Map;
import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import vn.edu.uit.careerhub.auth.AuthPrincipal;
import vn.edu.uit.careerhub.auth.RequestMetadata;
import vn.edu.uit.careerhub.common.AppException;
import vn.edu.uit.careerhub.placements.PlacementModels.Context;
import vn.edu.uit.careerhub.placements.PlacementModels.EvaluationRole;
import vn.edu.uit.careerhub.placements.PlacementModels.Status;
import vn.edu.uit.careerhub.placements.PlacementRequests.Evaluation;

@Service
public class EvaluationService {
    private final EvaluationRepository repository;

    public EvaluationService(EvaluationRepository repository) {
        this.repository = repository;
    }

    public Map<String, Object> get(AuthPrincipal principal, UUID applicationId, EvaluationRole role) {
        Context context = ownedContext(principal, applicationId, role, false);
        return repository.view(context, role);
    }

    @Transactional
    public Map<String, Object> submit(AuthPrincipal principal, UUID applicationId, UUID commandId,
            EvaluationRole role, Evaluation input, RequestMetadata request) {
        Context context = ownedContext(principal, applicationId, role, true);
        var repeated = repository.byCommand(context.placementId(), commandId);
        if (repeated.isPresent()) {
            if (!role.name().equals(repeated.get().get("respondentRole"))) {
                throw new AppException(HttpStatus.CONFLICT, "IDEMPOTENCY_KEY_REUSED",
                        "Idempotency-Key đã được dùng cho một phiếu khác.");
            }
            return repository.view(context, role);
        }

        Map<String, Object> current = repository.view(context, role);
        if (!Boolean.TRUE.equals(current.get("canSubmit"))) {
            if (context.status() != Status.COMPLETED) {
                throw new AppException(HttpStatus.CONFLICT, "INTERNSHIP_EVALUATION_NOT_AVAILABLE",
                        "Chỉ có thể gửi đánh giá sau khi UIT xác nhận hoàn thành kỳ thực tập.");
            }
            throw new AppException(HttpStatus.CONFLICT, "INTERNSHIP_EVALUATION_ALREADY_SUBMITTED",
                    "Phiếu đánh giá đã được gửi và không thể chỉnh sửa.");
        }

        repository.insert(context, role, principal.userId(), commandId, input, request);
        return repository.view(context, role);
    }

    private Context ownedContext(AuthPrincipal principal, UUID applicationId, EvaluationRole role, boolean lock) {
        Context context = repository.context(applicationId, lock)
                .orElseThrow(() -> new AppException(HttpStatus.NOT_FOUND, "INTERNSHIP_PLACEMENT_NOT_FOUND",
                        "Không tìm thấy kỳ thực tập."));
        boolean owned = role == EvaluationRole.STUDENT
                ? context.studentProfileId().equals(principal.studentProfileId())
                : context.companyId().equals(principal.companyId());
        if (!owned) {
            throw new AppException(HttpStatus.NOT_FOUND, "INTERNSHIP_PLACEMENT_NOT_FOUND",
                    "Không tìm thấy kỳ thực tập.");
        }
        return context;
    }
}
